core/src/main/java/org/bitcoinj/core/TransactionBag.java

```diff
@@ -49,4 +57,138 @@ public interface TransactionBag {
 
     /** Returns transactions from a specific pool. */
     Map<Sha256Hash, Transaction> getTransactionPool(WalletTransaction.Pool pool);
+
+    /**
+     * Returns false if this transaction has at least one output that is owned by the given wallet and unspent, true
+     * otherwise.
+     */
+    default boolean isEveryOwnedOutputSpent(Transaction tx) {
+        for (TransactionOutput output : tx.getOutputs()) {
+            if (output.isAvailableForSpending() && isMineOrWatched(output))
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * <p>Returns the list of transaction outputs, whether spent or unspent, that match a wallet by address or that are
+     * watched by a wallet, i.e., transaction outputs whose script's address is controlled by the wallet and transaction
+     * outputs whose script is watched by the wallet.</p>
+     *
+     * @param tx The transaction.
+     * @return linked list of outputs relevant to the wallet in this transaction
+     */
+    default List<TransactionOutput> getWalletOutputs(Transaction tx) {
+        List<TransactionOutput> walletOutputs = new LinkedList<>();
+        for (TransactionOutput o : tx.getOutputs()) {
+            if (!isMineOrWatched(o)) continue;
+            walletOutputs.add(o);
+        }
+
+        return walletOutputs;
+    }
+
+    /**
+     * Returns the difference of {@link #getValueSentToMe(Transaction)} and {@link #getValueSentFromMe(Transaction)}.
+     */
+    default Coin getValue(Transaction tx) throws ScriptException {
+        // TODO: Can we lose the commented-out caching code?
```

Comment:
The hack was added via ece8d9a3477c0cc517f4dbf36d37e1e9441725de in 2015. In general I'd say mobile devices are quick enough that special cases should not be needed. However I don't know about this specific case.

Reply:
See #2538.